rtengine/imageio.cc

```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
     return IMIO_SUCCESS;
 }
 
+#ifdef LIBJXL
+#define _PROFILE_ JXL_COLOR_PROFILE_TARGET_ORIGINAL
+// adapted from libjxl
+int ImageIO::loadJxl(const Glib::ustring &fname)
+{
+    if (pl) {
+        pl->setProgressStr("PROGRESSBAR_LOADJXL");
+        pl->setProgress(0.0);
+    }
+
+    std::vector<uint8_t> icc_profile;
+
+    gpointer buffer = nullptr;
+    size_t buffer_size = 0;
+
+    JxlBasicInfo info = {};
+    JxlPixelFormat format = {};
+
+    format.num_channels = 3;
+    format.data_type = JXL_TYPE_FLOAT;
+    format.endianness = JXL_NATIVE_ENDIAN;
+    format.align = 0;
+
+    // get file contents
+    std::ifstream instream(fname.c_str(), std::ios::in | std::ios::binary);
```

Comment:
How about `Gio::File` that is used in various places? But maybe @Lawrence37 has an idea on how problematic the `filename.c_str()` actually is...

Reply:
I found this, but I'm pretty sure there's a pointer to remember to free: `bool Gio::File::load_contents(char *&contents, gsize &length);`

Reply:
This seems to work... don't know if it's safe passing everything through std::string.
```C++
std::string const contents = Glib::file_get_contents(fname);
std::vector<std::uint8_t> const compressed(contents.begin(), contents.end());
```

Reply:
Hmm, `Glib::file_get_contents(fname)` takes a `std::string` not a `Glib::ustring` and also thows. Additionally, this doubles the amount of memory used. So no, this is not a solution, I'm afraid.

Reply:
Would be nice to have a function to read a file into a vector that could be used with all decoders that work with memory buffers.  Would such a function go in rtengine/utils.cc?  Maybe I should look...

I think I misunderstood your initial comment (again).  I didn't notice any issues reading files with Chinese and Japanese filenames using either fstream/c_str or file_get_contents.  I don't recall the reason for using c_str().  I vaguely remember having problems with casting or auto conversion to std::string not working.  Might be a glitch that's been resolved by now.

Reply:
`Glib::ustring::c_str()` returns a UTF-8 encoded Unicode string, but `std::ifstream`'s expected encoding is platform-dependent (I think?). Maybe [`Glib::filename_from_utf8`](https://developer-old.gnome.org/glibmm/2.48/group__CharsetConv.html#ga2bd94064ad97b43324a7854b62f0835b) can be used to ensure the correct encoding for the file name? I'm not sure if the encoding is meant for the platform, Gio, or both. For a full Gio method, you can try getting a file with [`Gio::File::create_for_path`](https://developer-old.gnome.org/glibmm/2.48/classGio_1_1File.html#a1b1a4553b226689d50b9525488223e25), opening an input stream with [`Gio::File::read`](https://developer-old.gnome.org/glibmm/2.48/classGio_1_1File.html#a0395b3b0b8cec5dc1fd9953e52e5fbc1), and finally read the file content into your buffer with [`Gio::InputStream::read_all`](https://developer-old.gnome.org/glibmm/2.48/classGio_1_1InputStream.html#a6da0fc57d1a5df7107370c67e84aa496).

Reply:
I added a function to `utils.cc` because it's potentially useful elsewhere.  It uses `Glib::filename_from_utf8` to convert ustring to std::string.  It returns an empty vector if there's any error.

```c++
std::vector<std::uint8_t> getFileData(const Glib::ustring &filename);
```